### ISSUE TITLE
emqttc is not sending PINGREQs to Mosquitto broker.

### ISSUE THREAD

**Issue by shymega** [emqtt/emqttc#51]
TL;DR: Mosquitto and emqttc do not work well with each other. Other brokers work fine with emqttc.

Erlang version: Erlang/OTP 20 [erts-9.0]
Elixir version: 1.5.1
Mosquitto version: 1.4.10

(Have also tested on Erlang/OTP 19.2 and Elixir 1.4.5)

The messages I get from Mosquitto are:
```
1505752794: Client xams_router has exceeded timeout, disconnecting.
1505752794: Socket error on client xams_router, disconnecting.
```

The log output from emqttc when the broker disconnects the client (This log uses a keepalive of 5 seconds):

```
19:06:30.903 [info]  MQTT client now connecting to broker.

19:06:30.903 [debug] Connecting without SSL.
[info] [Client <0.244.0>]: connecting to localhost:1883

19:06:30.915 [debug] Connected!
19:06:30.915 [info] Application xams started on node nonode@nohost
Interactive Elixir (1.4.5) - press Ctrl+C to exit (type h() ENTER for help)
iex(1)> [debug] [xams_router@127.0.0.1:33967] SENT: CONNECT(Q0, R0, D0, ClientId=xams_router, ProtoName=MQTT, ProtoVsn=4, CleanSess=false, KeepAlive=5, Username=xams, Password=******)
[debug] [xams_router@127.0.0.1:33967] SENT: <<16,39,0,4,77,81,84,84,4,192,0,5,
                                              0,11,120,97,109,115,95,114,111,
                                              117,116,101,114,0,4,120,97,109,
                                              115,0,8,112,97,115,115,119,48,
                                              114,100>>
[info] [Client <0.244.0>] connected with localhost:1883
[info] [Client <0.244.0>] RECV: CONNACK_ACCEPT

19:06:30.938 [info]  Success. MQTT client connected.
[debug] [xams_router@127.0.0.1:33967] SENT: SUBSCRIBE(Q1, R0, D0, PacketId=1, TopicTable=[{<<"xams/#">>,2}])

19:06:30.938 [debug] Begin subscribing phase.

19:06:30.938 [debug] Router about to subscribe to topic -> xams/#

19:06:30.938 [debug] Router now subscribed to topic -> xams/#

19:06:30.938 [debug] Finished subscribing phase.
[debug] [xams_router@127.0.0.1:33967] SENT: <<130,11,0,1,0,6,120,97,109,115,47,
                                              35,2>>
[debug] [Client <0.244.0>] RECV: SUBACK(Q0, R0, D0, PacketId=1, QosTable=[2])

nil
iex(2)> [warning] [Client <0.244.0>] Connection lost for: tcp_closed
```

The code used to start emqttc is:

```
:emqttc.start_link([
      host: host,
      port: port,
      client_id: client_id,
      clean_sess: false
      keepalive: 15,
      connack_timeout: 30,
      reconnect: {3, 60, 10},
      username: username,
      password: password])
```

The variables `client_id`, `host`, `port`, `username`, and `password` are all valid variables. They are not included here due to privacy concerns.

I subscribe to one topic - `xams/#` w/ the QoS setting set to `:qos2`.

I can't see any other reason for Mosquitto to do this. I've tested it on two different servers of mine, and the `iot.eclipse.org` Mosquitto server, which has the same issue.

EDIT:

I have tried the same code with VerneMQ and RabbitMQ. The bug is not present in those brokers.
I believe this issue is constrained to Mosquitto and emqttc.

**Comment by shymega** [comment]
Edited post. FWIW.

**Comment by shymega** [comment]
Update: tested it on VerneMQ. Issue is no longer present. Appears to be bug in Mosquitto.

**Comment by shymega** [comment]
I've concluded that this is a bug in Mosquitto. Not sure what to do next, so am idling.

**Comment by shymega** [comment]
I'd rather use Mosquitto, but I still can't figure out the issue per se. I think its to do with this library and Mosquitto in particular though. If any other information is required, please do let me know.

**Comment by shymega** [comment]
I have updated the issue to provide more information. Current debugging is not giving me much clues as to where the issue is,

**Comment by emqplus** [comment]
@shymega @HJianBo Please check if the `emqttc` sends the PINGREQs to Mosquitto timely.

**Comment by shymega** [comment]
@emqplus, @HJianBo Hi. 

Thanks for the response.

Do you require any extra information from my program in particular?

Thanks for your help.

**Comment by shymega** [comment]
I haven't heard anything for a while. I haven't been testing the Mosquitto broker lately, and I'm unclear on **how** to get the required diagnostics from `emqttc` and Mosquitto. I have Mosquitto's logging up to the highest it will go, so I would think to get more detailed data, I would have to run it through a debugger. ..

**Comment by HJianBo** [comment]
Hi @shymega sorry for the late reply!

The reason for the error is indeed the client did not send PINGREQ timely!

When the `emqttc` received others Control Packets(for example: SUBACK), the client think the Keepalive is correct. So the PINGREQs Packet will be sent at next check turn. But the time interval of twice sending action more than a Keepalive.

https://github.com/emqtt/emqttc/blob/d43c110f7f39a8faa3d38f781638639f01fc6cb0/src/emqttc_keepalive.erl#L86-L95

Ref: **MQTT-3.1.2-23**
> It is the responsibility of the Client to ensure that the interval between Control Packets being sent does not exceed the Keep Alive value.

**Comment by shymega** [comment]
Hello,

HeeeJianBo <notifications@github.com> writes:

> Hi @shymega sorry for the late reply!

No worries!

> The reason for the error is indeed the client did not send PINGREQ timely!
>
> When the `emqttc` received others Control Packets(for example:
> SUBACK), the client think the Keepalive is correct. So the PINGREQs
> Packet will be sent at next check turn. But the time interval of twice
> sending action more than a Keepalive.

Right, I see.

> https://github.com/emqtt/emqttc/blob/d43c110f7f39a8faa3d38f781638639f01fc6cb0/src/emqttc_keepalive.erl#L86-L95

OK, so am I right in thinking that this code you've linked to is a
regression in the emqttc library?

What is interesting is that other MQTT servers work.. its just
Mosquitto that has issues. Examples of other MQTT servers are VerneMQ
and RabbitMQ, which work fine..

--
Best regards,
Dom Rodriguez (shymega / dzr).

**Comment by HJianBo** [comment]
@shymega Yes, It will works fine maybe! If the server Keepalive checking range is more tolerant.

For example:
The **emqttd** server will conclude the connection having been broken, When not received any Control Packet after `2 to 3 times of keepalive * backoff`. 

(Not just one Keepalive time interval in here. And I guess Mosquitto may be one Keepalive interval only  )

So, The `emqttc` client can work well on emqttd broker!

**Comment by shymega** [comment]
@HJianBo I understand that emqttd works fine.

For my needs, using Mosquitto is pretty much the more sensible solution.. I don't exactly need scalability or durability for my MQTT broker -- in fact, its overkill.

I guess the main issue is to fix `emqttc`'s support with Mosquitto right now.

**Comment by shymega** [comment]
@HJianBo Is there any update on this? The bug is preventing further development of some of my projects, such as xams-project/xams, unless I was to test on a different broker.. I'd rather use Mosquitto though, for the lightweight footpoint.

**Comment by shymega** [comment]
@xsir99 -- what version of Mosquitto and Erlang (and/or Elixir?) are you using? Interested to know. I'm also assuming that you're using emqttc from git master?

Thanks :)

**Comment by shymega** [comment]
@emqplus Yeah, I would agree there -- a `force-ping` option would be a good interim solution until the bug can be properly fixed. I have the impression its a regression with Mosquitto itself, though.

I'd be happy to close this issue and raise it with the Mosquitto developers. Given that VerneMQ and other MQTT brokers work *fine* with emqttc, its clear that the issue does *not* lie with emqttc.

**Comment by xsir99** [comment]
@shymega   public broker test.mosquitto.org , emqttc tag 0.7.1-beta and  erlang otp 19.3 .
 It's a little different, the connection will works well, if you don't do anything just make a connect, however, after try to sub a topic,  it will be closed in a while

**Comment by shymega** [comment]
@xsir99 I see. I get a similar result from my own broker (also running Mosquitto). 

I think we'll have to see how the solution proposed by @emqplus works. If that doesn't resolve the issue, then further investigation may be required..

**Comment by emqplus** [comment]
@shymega I committed a fix on `force-ping` branch, pls verify the issue, thx.

**Comment by shymega** [comment]
@emqplus That works perfectly. I think further investigation is required, so perhaps we should keep this issue open and coordinate w/ Mosquitto devs about the bug?

For now, I think its a good interim solution, but fixing the root cause might be necessary. 

Thanks!

**Comment by shymega** [comment]
I also meant to add: as the fix works, will it be merged into `master`? Thanks.

**Comment by chintan-mishra** [comment]
I am also receiving `Connection lost for: tcp_closed` error. Even though KeepAlive for client < KeepAlive for my broker.

Erlang: 20.2
Elixir: 1.6.1

**Comment by emqplus** [comment]
@shymega I merged the PR #55 into `master` branch.

**Comment by shymega** [comment]
@emqplus As this has been fixed in `master`, I'm closing this issue. Thanks for your help.
